tests/rptest/tests/topic_delete_test.py

```diff
@@ -7,24 +7,108 @@
 # the Business Source License, use of this software will be governed
 # by the Apache License, Version 2.0
 
-from rptest.services.cluster import cluster
-from rptest.services.admin import Admin
-from ducktape.utils.util import wait_until
 import time
+import json
 
+from ducktape.utils.util import wait_until
+from ducktape.mark import parametrize
+from requests.exceptions import HTTPError
+
+from rptest.services.cluster import cluster
 from rptest.clients.types import TopicSpec
+from rptest.clients.offline_log_viewer import OfflineLogViewer
 from rptest.tests.redpanda_test import RedpandaTest
 from rptest.clients.kafka_cli_tools import KafkaCliTools
 from rptest.services.rpk_producer import RpkProducer
 from rptest.services.metrics_check import MetricCheck
 from rptest.services.redpanda import SISettings
 from rptest.util import wait_for_segments_removal
-from ducktape.mark import parametrize
+from rptest.services.admin import Admin
 
 
-def topic_storage_purged(redpanda, topic: str):
+def get_kvstore_topic_key_counts(redpanda):
```

Comment:
> Partitions have various kvstore items, stored by
the raft + storage layers.  When the topic is
deleted, they should be deleted too.

> Use our offline log decoder tool to inspect
the kvstore out-of-band and validate the keys
go away when the topic is deleted.

This is cool. What do you think the guidance should be on when to validate things on disk vs validating at a higher level? Applied too liberally it could eat up a lot of time and be hard to maintain.

For instance here what do we gain from reading the log compared to restarting redpanda and checking the kvstore? The reason being that we know kvstore serves data from memory and its only source of state is from the on disk file.

Reply:
When something is available via external APIs, I would generally expect tests to get it that way.

The first use case for this (in this PR) is for something we can't validate externally -- cleanup of data that is no longer needed.  It's similar to when we would use the existing storage inspection stuff: when we want to validate something beyond client-visible correctness, like storage being reclaimed, flushed up to a certain point, or migrated to/from tiered storage.

There are a bunch of storage tests that do things like "write 5 segments worth of data, wait til we see 5 segments, wait for 3 segments to upload to tiered storage...", which I anticipate being more robust if we convert them to use a decode of the log to instead assert on "wait til we see N batches, don't care how they're segmented".  Only for the tests that write very small amounts of data, of course.

I would also like to use this as a sanity check that log-structured things like the controller log are at a length we expect: right now it would be pretty easy to slip in a bug that caused like 1000 extra controller log messages (like some rogue fiber writing messages repeatedly), and nothing would notice/fail.  We could use some debug admin APIs for querying that kind of thing: the tradeoff is how much code we bake into redpanda proper vs. how much we can maintain as test-specific stuff.  Then again, stuff we bake into redpanda proper is more usable in field debug tools...

Long story short, I mostly anticipate doing direct on-disk reads in the context of storage tests.

Reply:
Thanks for the explanation

> I would also like to use this as a sanity check that log-structured things like the controller log are at a length we expect

i like it. i recall at some point @mmaslankaprv suggesting that we write a test that verified that all of the controller logs were identical (up to some common prefix) after a test run.  i don't know if that was ever done but this would make that a lot easier!